dali/dfu/dfuutil.cpp

```diff
@@ -857,21 +857,45 @@ class CDFUhelper: public CInterface, implements IDFUhelper
             else
                 throwError1(DFUERR_DFileNotFound, lfn);
         }
-        else {
-            Owned<IFileDescriptor> fdesc = file->getFileDescriptor();
-            Owned<IPropertyTree> t = fdesc->getFileTree();
+        else
+        {
+            Owned<IPropertyTree> t = queryDistributedFileDirectory().getFileTree(lfn, user, NULL, FOREIGN_DALI_TIMEOUT, true, true);
             toXML(t, out, true);
         }
         return out;
     }
 
-    void addFileXML(const char *lfn,const StringBuffer &xml,IUserDescriptor *user)
+    void addFileXML(const char *lfn, const StringBuffer &xml, IUserDescriptor *user)
     {
         Owned<IPropertyTree> t = createPTreeFromXMLString(xml);
-        Owned<IFileDescriptor> fdesc = deserializeFileDescriptorTree(t,&queryNamedGroupStore(),0);
-        Owned<IDistributedFile> file = queryDistributedFileDirectory().createNew(fdesc,true);
-        if (file)
-            file->attach(lfn,user);
+        // Check if this XML is a superfile map
+        if (t->hasProp("@numsubfiles"))
+        {
+            // It seems XML is fine to create  superfile
+            Owned<IDistributedSuperFile> superfile = queryDistributedFileDirectory().createSuperFile(t, user, false);
+        }
+        else
+        {
+            // Logical file map
+            if (queryDistributedFileDirectory().exists(lfn, user, true))
+                throw MakeStringException(-1, "Logical file '%s' already exists!", lfn);
+
+            Owned<IFileDescriptor> fdesc = deserializeFileDescriptorTree(t, &queryNamedGroupStore(), 0);
+            Owned<IDistributedFile> file = queryDistributedFileDirectory().createNew(fdesc, true);
+            if (file)
+            {
+                if (file->existsPhysicalPartFiles(0))
```

Comment:
Arguably this should be optional
There's no guarantee the physicals don't disappear after this command succeeds..
And there might be valid reasons for wanting to restore meta and physicals separately or not necessarily have to restore physicals ahead of meta.